src/main/java/com/amazon/opendistroforelasticsearch/sql/antlr/visitor/AntlrSqlParseTreeVisitor.java

```diff
@@ -217,11 +231,37 @@ public T visitUdfFunctionCall(UdfFunctionCallContext ctx) {
         return reduce(func, ctx.functionArgs());
     }
 
-    // This check should be able to accomplish in grammar
+    /**
+     * The scalar function calls are separated into (a)typical function calls; (b)nested function calls with functions
+     * as arguments, like abs(log(...)); (c)aggregations with functions as aggregators, like max(abs(....)).
+     * Currently, we do not support nested functions or nested aggregations, aka (b) and (c).
+     * However, for the special EsFunctions included in the [supportedNestedFunctions] set, we have supported them in
+     * nested function calls and aggregations (b&c). Besides, the math constants included in the [mathConstants] set
+     * are regraded as scalar functions, but they are working well in the painless script.
+     *
+     * Thus, the types of functions to throw exceptions:
+     * (I)case (b) except that the arguments are from the [mathConstants] set;
+     * (II) case (b) except that the arguments are from the [supportedNestedFunctions] set;
+     * (III) case (c) except that the aggregators are from thet [supportedNestedFunctions] set.
+     */
     @Override
     public T visitScalarFunctionCall(ScalarFunctionCallContext ctx) {
```

Comment:
Same suggestion for the new code you're adding here: could you move them to another class, ex. `UnsupportedFunction`?

Btw, could you explore the possibility of intercepting and checking what you're interested in an implementation class of `GenericSqlParseTreeVisitor` without touching grammar file? I think at least it's doable for unsupported function and operator which can help you minimize the changes.

Reply:
Ok will do!

Reply:
It's done now. Thanks!

Reply:
Thanks for the changes! And we can make it better. Sorry I didn't make it clear at first.

```
class AntlrSqlParseTreeVisitor {
    public T visitScalarFunctionCall(ScalarFunctionCallContext ctx) {
        xxx.verify(ctx);
        T func = visit(ctx.scalarFunctionName());
        return reduce(func, ctx.functionArgs());
    }

    public T visitMathOperator(MathOperatorContext ctx) {
        xxx.verify(ctx);
        return super.visitMathOperator(ctx);
    }
    ...
}

class XXX { // Need a good name for this abstraction (semantics that we don't support for now).
    private static final Set<String> mathConstants, unsupportedFunctions ...

    public void verify(ScalarFunctionCallContext ctx) {
        ...
    }

    public void verify(MathOperatorContext ctx) {
        if (unsupportedOperators.contains(StringUtils.toLower(ctx.getText())))
            throw new SqlFeatureNotImplementedException(...);
    }

   ...
}
```